src/hooks/useCancelablePromise.ts

```diff
@@ -0,0 +1,65 @@
+import { useState, useEffect, useRef } from "react";
+import {StateTuple, ResultTuple} from './usePromise'
+
+const CANCEL = Symbol('cancel')
+
+const createCancelablePromise = <T>(inputPromise: Promise<T>) => {
+  let cancel: Function;
+  const controller: Promise<void | typeof CANCEL> = new Promise((_, rej) => {
+    cancel = () => {
+      rej(CANCEL);
+    };
+  });
+
+  
+  const promise: Promise<T> = Promise.race([
+    inputPromise,
+    controller,
+  ]) as Promise<T>; 
+  
+  return {
+    promise,
+    cancel: cancel!, // TS do not know the callback of promise been executed, so use assertion
+  };
+};
+
+type PromiseCreator<T> = (...props: any[]) => Promise<T>;
+
+export default function useCancelablePromise<T>(
+  promiseCreator: PromiseCreator<T>,
+  ...props: any[]
```

Comment:
You may need this utility type: [`Parameters<Type>`](https://www.typescriptlang.org/docs/handbook/utility-types.html#parameterstype)